app/utils/entity_utils.rb

```diff
@@ -255,17 +306,36 @@ def transform_and_validate(fields, input)
   def define_builder(*specs)
     fields = parse_specs(specs)
 
-    -> (opts = {}) do
-      raise(TypeError, "Expecting an input hash. You gave: #{opts}") unless opts.is_a? Hash
+    -> (data, opts = {}) do
+      opts = {
+        specs_only: false,
+        result: false
+      }.merge(opts)
 
-      result = transform_and_validate(fields, opts)
+      return fields if opts[:specs_only]
 
-      unless (result[:errors].empty?)
-        loc = caller_locations(2, 1).first
-        raise(ArgumentError, "Error(s) in #{loc}: #{result[:errors]}")
-      end
+      raise(TypeError, "Expecting an input hash. You gave: #{data}") unless data.is_a? Hash
+
+      result = transform_and_validate(fields, data)
+
+      if !result[:errors].empty?
+        msg = result[:errors].map { |error|
+          "#{error[:field]}: #{error[:msg]}"
+        }.join(", ")
 
-      result[:value]
+        if opts[:result]
+          Result::Error.new(msg, result[:errors])
+        else
+          loc = caller_locations(2, 1).first
+          raise(ArgumentError, "Error(s) in #{loc}: #{msg}")
+        end
+      else
+        if opts[:result]
+          Result::Success.new(result[:value])
+        else
+          result[:value]
+        end
+      end
```

Comment:
Now that we have two different ways to run the tranformation and validation keeping the lambda as return value type makes the API a bit convoluted.

Instead, we could return an object which would have an API like:
```ruby
Shape = EntityUtils.define_builder(...);

# This is the default behavior, raise error if validation fails
Shape.build({...})

# This is the new behavior, return Result object
Shape.validate({...})

# This calls build and, afaik, should keep the old call syntax Shape[...] working too.
Shape.call({..})

Shape.specs # attr_reader to get the parsed specs
```